producer/writer/router.go

```diff
@@ -20,7 +20,54 @@
 
 package writer
 
+import (
+	"fmt"
+	"sync"
+)
+
 type ackRouter interface {
 	// Ack acks the metadata.
 	Ack(ack metadata) error
+
+	// Register registers a message writer.
+	Register(replicatedShardID uint64, mw messageWriter)
+
+	// Unregister removes a message writer.
+	Unregister(replicatedShardID uint64)
+}
+
+type router struct {
+	sync.RWMutex
+
+	mws map[uint64]messageWriter
+}
+
+func newAckRouter(size int) ackRouter {
+	return &router{
+		mws: make(map[uint64]messageWriter, size),
+	}
+}
+
+func (r *router) Ack(meta metadata) error {
+	r.RLock()
+	mw, ok := r.mws[meta.shard]
+	r.RUnlock()
+	if !ok {
+		// Unexpected.
+		return fmt.Errorf("can't find shard %v", meta.shard)
```

Comment:
Do you need a metric here?

Reply:
There is a metric on the caller side of this method